jira/resilientsession.py

```diff
@@ -1,119 +1,288 @@
+import abc
 import json
 import logging
 import random
 import time
-from typing import Callable, Optional, Union, cast
+from typing import Any, Dict, Optional, Union
 
 from requests import Response, Session
 from requests.exceptions import ConnectionError
+from requests.structures import CaseInsensitiveDict
+from typing_extensions import TypeGuard
 
 from jira.exceptions import JIRAError
 
-logging.getLogger("jira").addHandler(logging.NullHandler())
+LOG = logging.getLogger(__name__)
 
 
-def raise_on_error(r: Optional[Response], verb="???", **kwargs):
+class PrepareRequestForRetry(metaclass=abc.ABCMeta):
+    """This class allows for the manipulation of the Request keyword arguments before a retry.
+
+    The :py:meth:`.prepare` handles the processing of the Request keyword arguments.
+    """
+
+    @abc.abstractmethod
+    def prepare(
+        self, original_request_kwargs: CaseInsensitiveDict
+    ) -> CaseInsensitiveDict:
+        """Process the Request's keyword arguments before retrying the Request.
+
+        Args:
+            original_request_kwargs (CaseInsensitiveDict): The keyword arguments of the Request.
+
+        Returns:
+            CaseInsensitiveDict: The new keyword arguments to use in the retried Request.
+        """
+        return original_request_kwargs
+
+
+class PassthroughRetryPrepare(PrepareRequestForRetry):
+    """Returns the Request's keyword arguments unchanged, when no change needs to be made before a retry."""
+
+    def prepare(
+        self, original_request_kwargs: CaseInsensitiveDict
+    ) -> CaseInsensitiveDict:
+        return super().prepare(original_request_kwargs)
+
+
+def raise_on_error(resp: Optional[Response], **kwargs) -> TypeGuard[Response]:
     """Handle errors from a Jira Request
 
     Args:
-        r (Optional[Response]): Response from Jira request
-        verb (Optional[str]): Request type, e.g. POST. Defaults to "???".
+        resp (Optional[Response]): Response from Jira request
 
     Raises:
         JIRAError: If Response is None
         JIRAError: for unhandled 400 status codes.
-        JIRAError: for unhandled 200 status codes.
+
+    Returns:
+        TypeGuard[Response]: True if the passed in Response is all good.
     """
     request = kwargs.get("request", None)
-    # headers = kwargs.get('headers', None)
 
-    if r is None:
+    if resp is None:
         raise JIRAError(None, **kwargs)
```

Comment:
since you're touching this part ... maybe make the exception slighlty more elaborate